### DISCUSSION TITLE
Enhancement Request for PowerOnState

### DISCUSSION THREAD

**Discussion by ralphrmartin** [arendst/Tasmota#11854] [Q&A] [answered]
### PROBLEM DESCRIPTION
PowerOnState options are too limiting.
 In the case of an extended power cut for example, it is not appropriate to just switch the relay on, or off, or even to set it to its previous state. E.g. we may only want a lamp to come on at night.
In cases where the relay is controlled by timers, it would be much more appropriate to (a) check the current time at power on, (b) set the relay to whatever the timers imply, given the current time of day.
Could this be added as an option for some future Tasmota release?

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): Sonoff S26, but applies to any controller
- [X] Tasmota binary firmware version number used: 9.4.0
  - [X] Pre-compiled
  - [ ] Self-compiled
- [X] Flashing tools used: OTA
- [X] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
06:40:51.799 CMD: Backlog Template; Module; GPIO 255
06:40:51.822 RSL: RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
06:40:52.044 RSL: RESULT = {"Module":{"8":"Sonoff S2X"}}
06:40:52.299 RSL: RESULT = {"GPIO0":{"32":"Button1"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"224":"Relay1"},"GPIO13":{"320":"Led_i1"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}

```
- [X] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
Not using rules

```
- [X] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
06:42:10.680 CMD: Status 0
06:42:10.687 RSL: STATUS = {"Status":{"Module":8,"DeviceName":"Tasmota_Right_Propagator","FriendlyName":["Tasmota_Right_Propagator"],"Topic":"tasmota_1C5B1C","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
06:42:10.693 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T08:50:06","StartupUTC":"2021-04-22T20:52:04","Sleep":50,"CfgHolder":4617,"BootCount":9,"BCResetTime":"2021-04-08T13:40:27","SaveCount":63,"SaveAddress":"F9000"}}
06:42:10.697 RSL: STATUS2 = {"StatusFWR":{"Version":"9.4.0(tasmota)","BuildDateTime":"2021-04-21T12:48:39","Boot":7,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"421/699"}}
06:42:10.701 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["rrmair",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C8000100060000005A0A000000000000","00000081","00006000","00000000"]}}
06:42:10.710 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":598,"Free":404,"Heap":26,"ProgramFlashSize":1024,"FlashSize":1024,"FlashChipId":"146085","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6"}}
06:42:10.715 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota_1C5B1C-6940","IPAddress":"192.168.1.137","Gateway":"192.168.1.254","Subnetmask":"255.255.255.0","DNSServer":"192.168.1.102","Mac":"CC:50:E3:1C:5B:1C","Webserver":2,"WifiConfig":4,"WifiPower":17.0}}
06:42:10.719 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_1C5B1C","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
06:42:10.725 RSL: STATUS7 = {"StatusTIM":{"UTC":"2021-04-23T05:42:10","Local":"2021-04-23T06:42:10","StartDST":"2021-03-28T01:00:00","EndDST":"2021-10-31T02:00:00","Timezone":99,"Sunrise":"06:00","Sunset":"20:30"}}
06:42:10.730 RSL: STATUS10 = {"StatusSNS":{"Time":"2021-04-23T06:42:10"}}
06:42:10.735 RSL: STATUS11 = {"StatusSTS":{"Time":"2021-04-23T06:42:10","Uptime":"0T08:50:06","UptimeSec":31806,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"ON","Wifi":{"AP":1,"SSId":"rrmair","BSSId":"18:E8:29:CA:34:26","Channel":1,"RSSI":56,"Signal":-72,"LinkCount":1,"Downtime":"0T00:00:03"}}}
06:42:13.450 RSL: STATE = {"Time":"2021-04-23T06:42:13","Uptime":"0T08:50:09","UptimeSec":31809,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"ON","Wifi":{"AP":1,"SSId":"rrmair","BSSId":"18:E8:29:CA:34:26","Channel":1,"RSSI":56,"Signal":-72,"LinkCount":1,"Downtime":"0T00:00:03"}}

```
- [X] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
Not an issue which is triggered as such

```

### TO REPRODUCE
Use any Tasmota device, power cycle it.

### EXPECTED BEHAVIOUR
Ideal behaviour would be to set relay according to time of day, and timers, not just set relay to off, on, or previous state.

### SCREENSHOTS
Not needed.

### ADDITIONAL CONTEXT

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by ascillato2** [comment, marked as answer]
Your desired behaviour can be accomplished now by the use of rules.

**Reply by ralphrmartin** [reply]
If you could give me a pointer, I'd be grateful - thanks. 

But this seems like an obvious requirement many people will have for default behaviour for anything that is controlled by timers, which is why I hope it could be one of the options for PowerOnState, without having to learn about rules.

**Reply by barbudor** [reply]
https://tasmota.github.io/docs/Rules/ is the place to start

What you are asking is unlikely to be achievable as no system can easily guessed what should be it's state when it comes back on line after a power-cut as it is missing all the history.
If you timer schedule is simple, it can easily be replaced by a rule which would work at restart.
If not, you may be able to manage that on an external back-end that would run out an UPS and know what state should be restored.

**Reply by ralphrmartin** [reply]
To help others trying to find it, a good place to start at https://tasmota.github.io/docs/Rules/ is "Make sure Light is on at night".

Clearly, a general solution is not feasible. But I suspect many people use their Tasmota devices as time switches, based around very simple logic of the kind 
Every day, for a list of times T1, T2,... alternately switch the device on and off (knowing whether it starts the day on or off). The times might be actual times, or sunset/sunrise.

I am suggesting in cases where the Tasmota device is used in this very simple way, having PowerOnState determined by the timer control list is a straightforward and useful thing to do,  which will save many people from the complexity of rules. 

These devices are meant to make user's lives simpler, right? Not more complicated. :-)

**Reply by ascillato** [reply]
There you have a nice idea.

At this moment, Tasmota timers are just one-shot trigger timers, meaning that they will trigger ONLY in the exact moment you have configured.

What you are proposing are interval timers that will trigger IF the time is between the interval.

May be easier to understand and implement if they are added in the Timers menu rather than with an extra console command.

This was requested long time ago, but nobody spend time working on this. Do you want to work on this?

**Reply by ralphrmartin** [reply]
I'm afraid I'm rather too busy to do the work, but I'd be happy to help review and test a proposed solution.

**Reply by barbudor** [reply]
I subscribe to @ascillato , if you can come with PR which implements that into a nice and memory efficient way, your PR will be welcomed

**Reply by ascillato** [reply]
Just for reference, below are links to some of the old issues where this was discussed:

https://github.com/arendst/Tasmota/issues/8643

https://github.com/arendst/Tasmota/issues/3475

**Reply by ralphrmartin** [reply]
A very straightforward way to implement this (for a daily timer, same idea works for weeks, etc) would be to have 
(1) a 1 bit flag which says whether the output is be initially off or on at time 00:00
(2) a list of times at which the output is alternately toggled on and off

On start up, the device gets the current time, and counts parity (on,off,on,off...) along the list until it it finds a time later than the current time.
The last parity change tells the device what the power on state should be initialised to.
An interrupt is also set to occur at that later time. 
When the interrupt occurs, it toggles the output, reads the next time from the list, and sets an interrupt for that time.
This repeats until you get to the end of the list, when you go back to the start of the list for the next day.

So, these are still one shot timers, but with correct state initialisation when the device powers up.
